File: 208. Implement Trie (Prefix Tree)/Solution.java
class Trie {
    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode curr = root;
        char[] arr = word.toCharArray();
        for (char c : arr) {
            if (curr.neighbours[c - 'a'] == null) curr.neighbours[c - 'a'] = new TrieNode();
            curr = curr.neighbours[c - 'a'];
        }
        curr.ends.add(word);
    }

    public TrieNode presearch(String s) {
        TrieNode curr = root;
        char[] arr = s.toCharArray();
        for (char c : arr) {
            if (curr.neighbours[c - 'a'] == null) return null;
            curr = curr.neighbours[c - 'a'];
        }
        return curr;
    }
    
    public boolean search(String word) {
        TrieNode curr = presearch(word);
        return curr == null ? false : curr.ends.contains(word);
    }
    
    public boolean startsWith(String prefix) {
        return presearch(prefix) != null;
    }
}

class TrieNode {
    TrieNode[] neighbours;
    Set<String> ends;
    public TrieNode() {
        neighbours = new TrieNode[26];
        ends = new HashSet<>();
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208. Implement Trie (Prefix Tree)/Solution.py
class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            if not curr.neighbours[ord(c) - 97]: curr.neighbours[ord(c) - 97] = TrieNode()
            curr = curr.neighbours[ord(c) - 97]
        curr.ends.add(word)

    def presearch(self, s):
        curr = self.root
        for c in s:
            if not curr.neighbours[ord(c) - 97]: return None
            curr = curr.neighbours[ord(c) - 97]
        return curr

    def search(self, word: str) -> bool:
        curr =  self.presearch(word)
        return False if not curr else word in curr.ends

    def startsWith(self, prefix: str) -> bool:
        return bool(self.presearch(prefix))
        
class TrieNode:
    def __init__(self):
        self.neighbours = [None for i in range(26)]
        self.ends = set()

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)